Source/WebKit/Platform/IPC/glib/ConnectionGLib.cpp

```diff
@@ -0,0 +1,576 @@
+                RELEASE_ASSERT(!m_incomingHardwareBuffers.isEmpty());
+                attachments[attachmentIndex] = WTFMove(m_incomingHardwareBuffers.first());
+                m_incomingHardwareBuffers.removeAt(0);
+            }
+            break;
+        case AttachmentInfo::Type::Unset:
+            RELEASE_ASSERT_NOT_REACHED();
+        }
+#else
+        if (!attachmentInfo.isNull())
+            attachments[attachmentIndex] = WTFMove(m_fileDescriptors[fdIndex++]);
+#endif
+    }
+
+    if (!messageInfo.isBodyOutOfLine())
+        return Decoder::create(messageData.first(messageInfo.bodySize()), WTFMove(attachments));
+
+    ASSERT(messageInfo.bodySize());
+    auto& attachmentInfo = reinterpretCastSpanStartTo<AttachmentInfo>(messageData);
+    if (attachmentInfo.isNull()) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    auto handle = WebCore::SharedMemory::Handle { WTFMove(m_fileDescriptors[fdIndex]), messageInfo.bodySize() };
+    auto messageBody = WebCore::SharedMemory::map(WTFMove(handle), WebCore::SharedMemory::Protection::ReadOnly);
+    if (!messageBody) {
+        ASSERT_NOT_REACHED();
+        return nullptr;
+    }
+
+    return Decoder::create(messageBody->mutableSpan().first(messageInfo.bodySize()), WTFMove(attachments));
+}
+
+static ssize_t readBytesFromSocket(GSocket* socket, Vector<uint8_t>& buffer, Vector<UnixFileDescriptor>& fileDescriptors, GCancellable* cancellable, GError** error)
+{
+    GUniqueOutPtr<GSocketControlMessage*> messages;
+    gint messagesCount = 0;
+    gint flags = 0;
+    GInputVector inputVector = { buffer.mutableSpan().data(), buffer.size() };
+    auto bytesRead = g_socket_receive_message(socket, nullptr, &inputVector, 1, &messages.outPtr(), &messagesCount, &flags, cancellable, error);
+    if (bytesRead <= 0)
+        return bytesRead;
+
+    if (flags & MSG_CTRUNC) {
+        // Control data has been discarded, so consider this a read failure.
+        return -1;
+    }
+
+    buffer.shrink(bytesRead);
+    for (int i = 0; i < messagesCount; ++i) {
+        GRefPtr<GSocketControlMessage> controlMessage = adoptGRef(G_SOCKET_CONTROL_MESSAGE(messages.get()[i]));
+        if (!G_IS_UNIX_FD_MESSAGE(controlMessage.get()))
+            continue;
+
+        gint fdsCount;
+        GUniquePtr<gint> fds(g_unix_fd_message_steal_fds(G_UNIX_FD_MESSAGE(controlMessage.get()), &fdsCount));
+        for (int i = 0; i < fdsCount; ++i) {
+            int fd = fds.get()[i];
+            if (!setCloseOnExec(fd)) {
+                ASSERT_NOT_REACHED();
+                break;
+            }
+
+            fileDescriptors.append(UnixFileDescriptor { fd, UnixFileDescriptor::Adopt });
+        }
+    }
+
+    return bytesRead;
+}
+
+void Connection::readyReadHandler()
+{
+#if OS(ANDROID)
+    if (m_pendingIncomingHardwareBufferCount) {
+        if (!receiveIncomingHardwareBuffers())
+            return;
+
+        if (auto decoder = createMessageDecoder(m_readBuffer, m_fileDescriptors))
```

Comment:
This method does not take any parameters:

```suggestion
        if (auto decoder = createMessageDecoder())
```